code/game/mecha/mech_prosthetics.dm

```diff
@@ -0,0 +1,365 @@
+	if(current)
+		data["current"] = current.name
+	data["queue"] = get_queue_names()
+	data["buildable"] = get_build_options()
+	data["category"] = category
+	data["categories"] = categories
+	if(all_robolimbs)
+		var/list/T = list()
+		for(var/A in all_robolimbs)
+			var/datum/robolimb/R = all_robolimbs[A]
+			if(R.unavailable_to_build) continue
+			T += list(list("id" = A, "company" = R.company))
+		data["manufacturers"] = T
+		data["manufacturer"] = manufacturer
+	data["materials"] = get_materials()
+	data["maxres"] = res_max_amount
+	data["sync"] = sync_message
+	if(current)
+		data["builtperc"] = round((progress / current.time) * 100)
+
+	ui = nanomanager.try_update_ui(user, src, ui_key, ui, data, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "mechfab.tmpl", "Prosthetics Fab UI", 800, 600)
+		ui.set_initial_data(data)
+		ui.open()
+		ui.set_auto_update(1)
+
+/obj/machinery/pros_fabricator/Topic(href, href_list)
+	if(..())
+		return
+
+	if(href_list["build"])
+		add_to_queue(text2num(href_list["build"]))
+
+	if(href_list["remove"])
+		remove_from_queue(text2num(href_list["remove"]))
+
+	if(href_list["category"])
+		if(href_list["category"] in categories)
+			category = href_list["category"]
+
+	if(href_list["manufacturer"])
+		if(href_list["manufacturer"] in all_robolimbs)
+			manufacturer = href_list["manufacturer"]
+
+	if(href_list["eject"])
+		eject_materials(href_list["eject"], text2num(href_list["amount"]))
+
+	if(href_list["sync"])
+		sync()
+	else
+		sync_message = ""
+
+	return 1
+
+/obj/machinery/pros_fabricator/attackby(var/obj/item/I, var/mob/user)
+	if(busy)
+		user << "<span class='notice'>\The [src] is busy. Please wait for completion of previous operation.</span>"
+		return 1
+	if(default_deconstruction_screwdriver(user, I))
+		return
+	if(default_deconstruction_crowbar(user, I))
+		return
+	if(default_part_replacement(user, I))
+		return
+
+	if(istype(I,/obj/item/weapon/disk/limb))
+		var/obj/item/weapon/disk/limb/D = I
+		if(!D.company || !(D.company in all_robolimbs))
+			user << "<span class='warning'>This disk seems to be corrupted!</span>"
+		else
+			user << "<span class='notice'>Installing blueprint files for [D.company]...</span>"
+			if(do_after(user,50,src))
+				var/datum/robolimb/R = all_robolimbs[D.company]
+				R.unavailable_to_build = 0
+				user << "<span class='notice'>Installed [D.company] blueprints!</span>"
+				qdel(I)
+		return
+
+	var/material
+	switch(I.type)
+		if(/obj/item/stack/material/gold)
+			material = "gold"
+		if(/obj/item/stack/material/silver)
+			material = "silver"
+		if(/obj/item/stack/material/diamond)
+			material = "diamond"
+		if(/obj/item/stack/material/phoron)
+			material = "phoron"
+		if(/obj/item/stack/material/steel)
+			material = DEFAULT_WALL_MATERIAL
+		if(/obj/item/stack/material/glass)
+			material = "glass"
+		if(/obj/item/stack/material/uranium)
+			material = "uranium"
+		if(/obj/item/stack/material/plasteel)
+			material = "plasteel"
```

Comment:
maaaaateeeerial systeeeeeeem

use the vars on the stack to get its material directly, not a huge switch :o